### ISSUE TITLE
Limit download/upload speeds (throttling)

### ISSUE THREAD

**Issue by green13** [webtorrent/webtorrent#163]
It would be cool to be able to limit download/upload speeds.

**Comment by servermeta** [comment]
As a noobish system administrator it feels a bit strange to me to use Javascript to limit bandwidth. Maybe it's just a prejudice, but node doesn't strike me as being a good candidate.

If it's ok to limit the solution to only linux systems, we could use Ip/Nftables and tc. We could open the stream as a child process, and then the linux kernel could keep track of the process, which is very efficient at doing it. 


```
PID=`ps aux |grep inetd |head -n 1 |cut -b 10-14`
```

First we get the PID of the process

```
/usr/local/sbin/iptables -A OUTPUT -p TCP -m owner --pid-owner $PID -j ACCEPT
````

we could use iptables to mark the packet, then Traffic control could shaper on a per PID basis

http://lartc.org/howto/lartc.qdisc.filters.html

**Comment by Svenskunganka** [comment]
@servermeta It's bad habit to rely on binaries outside the project, especially those who require root privileges. Node is definitely more than qualified for this task, it is known for its I/O and networking capabilities, we should utilize that.

**Comment by servermeta** [comment]
@Svenskunganka the only thing we would have to rely on is a unix kernel :P

**Comment by astro** [comment]
No, iptables is specific to Linux.

**Comment by servermeta** [comment]
yes you are right. I'm studying the source to try to come up with a good solution, since I need this, and to better learn this project. @feross why you say 
`(Btw, it doesn't make much sense for the browser version of webtorrent).`

**Comment by astro** [comment]
While dropping packets is perfectly valid, this shaping is best done in the actual code.

FWIW here's my approach from Bitford: https://github.com/astro/bitford/blob/master/src/shaper.js
I bet there are similar modules on npmjs.org.

**Comment by bradisbell** [comment]
Once implemented, will stream backpressure on the read side slow the download speed?

**Comment by astro** [comment]
Backpressure should go into https://github.com/feross/bittorrent-protocol/ where it doesn't seem to be implemented at this time. That will only become a problem if download speeds exceed CPU power & disk throughput.

**Comment by bradisbell** [comment]
@astro In the download case yes, but when streaming media it is very common that the download speed greatly exceed the playback speed.  For example, if you pipe media into VLC, only a couple seconds of media data will be buffered prior to decoding and playback.

**Comment by feross** [comment]
https://github.com/substack/node-brake

**Comment by mbrownnycnyc** [comment]
What happened to this issue?

**Comment by Deathspike** [comment]
I'm guessing this didn't happen yet?

The stream solution that you provided @feross is only for download streams, right?

It's no way to control upload. I'm pretty new in webtorrent, is any of the libs to seed providing support for throttling?

**Comment by zvxy** [comment]
:+1: 

Implementing this seems hard with the way the event loop is ran. It seems to walk the swarm a lot.

**Comment by scotttrinh** [comment]
I'm working on something similar at work (my application just uploads) and this library works really well: https://github.com/tjgq/node-stream-throttle

You can set up a group with an aggregate rate limit and spin off throttles from each. I'm working out issues with changing the speed while connections are in-flight, but if I get it sorted, I'd be happy to take a stab at a PR for this.

**Comment by feross** [comment]
@scotttrinh Yes, that would be much appreciated!

**Comment by coleplx** [comment]
Any update on this?
It would be great to finally be able to limit users connection :)

**Comment by scotttrinh** [comment]
@coleplx nothing on my end. I wasn't able to get changing the Throttle mid-stream to work for me, but since Webtorrent deals with smaller chunks, maybe it's acceptable to finish the in-flight chunks at the old speed? @feross ? If that's cool, I can put together a PR.

**Comment by scotttrinh** [comment]
Actually, I just saw #1040. That seems like a reasonable PR for this feature.

**Comment by abcd9999** [comment]
Hello, I would like to know if this is implemented, and if it is, then how to do it.

**Comment by scotttrinh** [comment]
@abcd9999 Unfortunately I never got around to it. :(

**Comment by quicoto** [comment]
This seems to have a very low priority although it would have a great impact.

**Comment by farlock85** [comment]
@feross could PR #1040 be re-evaluated again? If there is an issue with it I can have a look at it again.

**Comment by craigharman** [comment]
It's been 6 years now, assume this isn't going to be done?

**Comment by alxhotel** [comment]
@craigharman there is a PR pending  #1762. If you want you want test it and help getting the PR merged

**Comment by craigharman** [comment]
> @craigharman there is a PR pending #1762. If you want you can test it and help getting the PR merged

Sure @alxhotel , functionally works really well. Just needs the tests to work in regards to seeding and some modifications to the opts section of the docs API. I have commented this in review.

**Comment by subins2000** [comment]
There's currently no pause function to close all the active connections of a torrent and actually pause. From #543, this would also help implement the "real pausing" I infer.

**Comment by drakonkat** [comment]
If this has released now how is possible to limit download/upload speeds? i cannot find nothing in documentation

**Comment by codeit-ninja** [comment]
I dont understand, this issue was fixed in v1.3.0, currently I am on version 1.8.6 and there is no option in the client to set the download and or upload speed. Am I missing something?

![image](https://user-images.githubusercontent.com/22219122/161466830-60e141b3-2b6f-4453-9e59-29983c5e7129.png)

**Comment by scotttrinh** [comment]
@codeit-ninja 

Looks like no one has updated the DefinitelyTyped TypeScript definitions. You'll need to either use some kind of ambient declaration to add the missing properties, or better yet, submit a PR to add these missing keys and methods.
